File: SingleInstance/app/src/main/java/com/longing/singleinstance/onjava/LazyThreadSafeDoubleCheck.java
package com.longing.singleinstance.onjava;

/**
 * 4双重校验锁
 */
public class LazyThreadSafeDoubleCheck {
    private static volatile LazyThreadSafeDoubleCheck INSTANCE;

    private LazyThreadSafeDoubleCheck() {
    }

    public LazyThreadSafeDoubleCheck getInstance() {
        if (INSTANCE == null) {
            synchronized (LazyThreadSafeDoubleCheck.class) {
                if (INSTANCE == null) {
                    INSTANCE = new LazyThreadSafeDoubleCheck();
                }
            }
        }
        return INSTANCE;
    }
}

File: SingleInstance/app/src/main/java/com/longing/singleinstance/onkotlin/LazyThreadSafeDoubleCheck.kt
package com.longing.singleinstance.onkotlin

/**
 * 双重校验
 */
class LazyThreadSafeDoubleCheck private constructor() {
    companion object {
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            LazyThreadSafeDoubleCheck()
        }
    }

}